src/app/src/__tests__/sensorutils.tests.js

```diff
@@ -0,0 +1,134 @@
+                noDataValue: -999999,
+            },
+        };
+    };
+
+    const testSensorData = times(4, createTestSensorReading);
+
+    const expectedParsedSensorData = {
+        id: '0',
+        TEMPERATURE: 1,
+        OXYGEN: 1,
+        PH: 1,
+        TURBIDITY: 1,
+    };
+
+    const actualParsedSensorData = parseRiverGaugeApiData('0', testSensorData);
+
+    expect(actualParsedSensorData).toEqual(expectedParsedSensorData);
+});
+
+it('parses quarterly sensor data', () => {
+    const testSensorData = [
+        {},
+        {
+            p00010: 5.2,
+            p00300: 12.4,
+            p00400: 7.3,
+            p63680: 3.3,
+            sample_dt: '2018-12-04',
+            sample_start_time_datum_cd: 'EST',
+            sample_tm: '12:00',
+            site_no: 1438500,
+        },
+    ];
+
+    const actualParsedSensorData = parseRiverGaugeCsvData('0', testSensorData);
+
+    const expectedParsedSensorData = {
+        id: '0',
+        TEMPERATURE: 5.2,
+        OXYGEN: 12.4,
+        PH: 7.3,
+        TURBIDITY: 3.3,
+    };
+
+    expect(actualParsedSensorData).toEqual(expectedParsedSensorData);
+});
+
+it('calculates an overall sensor rating', () => {
+    const expectedGoodSensorRatings = {
+        TEMPERATURE: VARIABLE_WITHIN_HEALTHY_RANGE,
+        OXYGEN: VARIABLE_WITHIN_HEALTHY_RANGE,
+        PH: VARIABLE_WITHIN_HEALTHY_RANGE,
+        TURBIDITY: VARIABLE_WITHIN_HEALTHY_RANGE,
+    };
+
+    const expectedFairSensorRatings = {
+        TEMPERATURE: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        OXYGEN: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        PH: VARIABLE_WITHIN_HEALTHY_RANGE,
+        TURBIDITY: VARIABLE_WITHIN_HEALTHY_RANGE,
+    };
+
+    const expectedPoorSensorRatings = {
+        TEMPERATURE: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        OXYGEN: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        PH: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        TURBIDITY: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+    };
+
+    expect(RATING_GOOD).toEqual(
+        calculateOverallSensorRating(expectedGoodSensorRatings)
+    );
+    expect(RATING_FAIR).toEqual(
+        calculateOverallSensorRating(expectedFairSensorRatings)
+    );
+    expect(RATING_POOR).toEqual(
+        calculateOverallSensorRating(expectedPoorSensorRatings)
+    );
+});
+
+it('transforms sensor data to sensor ratings', () => {
+    const testParsedSensorData = {
+        id: '01438500',
+        TEMPERATURE: 5.2,
+        OXYGEN: 12.4,
+        PH: 7.3,
+        TURBIDITY: 3.3,
+    };
+
+    const expectedSensorRatings = {
+        TEMPERATURE: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        OXYGEN: VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        PH: VARIABLE_WITHIN_HEALTHY_RANGE,
+        TURBIDITY: VARIABLE_WITHIN_HEALTHY_RANGE,
+        OVERALL_RATING: RATING_FAIR,
+    };
+
+    const actualSensorRatings = transformSensorDataToRatings(
+        testParsedSensorData
+    );
+
+    expect(actualSensorRatings.sensorRatings).toEqual(expectedSensorRatings);
```

Comment:
Do we know whether `toEqual` will fully compare the object's keys and its values for equality? I've been using lodash `isEqual(x, y).toBe(true)` for similar checks, but it may be that `toEqual` is already making the proper comparison.

Reply:
In my limited experience, it has proven to be surprisingly robust for comparing objects. It has reported `false` when a value has differed a few keys down.

Reply:
Cool, I'll try it out next time I'm writing a test.

Reply:
Test-driven development was helpful here. Thanks for the encouragement.